crates/llm-chain-sagemaker-endpoint/src/model.rs

```diff
@@ -0,0 +1,176 @@
+use aws_sdk_sagemakerruntime::operation::invoke_endpoint::InvokeEndpointOutput;
+use aws_sdk_sagemakerruntime::primitives::Blob;
+use llm_chain::options::{ModelRef, Opt, OptDiscriminants, OptionsCascade};
+use llm_chain::prompt::Prompt;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use serde_with::skip_serializing_none;
+use strum_macros::EnumString;
+
+/// The `Model` enum represents the available SageMaker Endpoint models.
+/// Use SageMaker JumpStart to deploy the model listed here. Or use Model::Other
+/// to reference your custom models. For Model::Other, you need to write your own
+/// formatting logic for the request and response.
+///
+/// # Example
+///
+/// ```
+/// use llm_chain_sagemaker_endpoint::model::Model;
+///
+/// let falcon_model = Model::Falcon7BInstruct;
+/// let custom_model = Model::Other("your_custom_model_name".to_string());
+/// ```
+#[derive(Debug, Default, Clone, Serialize, Deserialize, EnumString, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum Model {
+    #[default]
+    #[strum(serialize = "falcon-7b-instruct")]
+    Falcon7BInstruct,
+
+    #[strum(serialize = "falcon-40b-instruct")]
+    Falcon40BInstruct,
+
+    /// A variant that allows you to specify a custom model name as a string, in case new models
+    /// are introduced or you have access to specialized models.
+    #[strum(default)]
+    Other(String),
+}
+
+pub trait Formatter {
+    fn format_request(&self, prompt: &Prompt, options: &OptionsCascade) -> Blob;
+    fn request_content_type(&self) -> String;
+    fn parse_response(&self, response: InvokeEndpointOutput) -> String;
+}
+
+impl Formatter for Model {
+    fn format_request(&self, prompt: &Prompt, options: &OptionsCascade) -> Blob {
+        match self {
+            Model::Falcon7BInstruct | Model::Falcon40BInstruct => {
+                #[skip_serializing_none]
+                #[derive(Serialize)]
+                struct Parameters {
+                    max_new_tokens: Option<usize>,
+                    max_length: Option<usize>,
+                    temperature: Option<f32>,
+                    top_k: Option<i32>,
+                    top_p: Option<f32>,
+                    stop: Option<Vec<String>>,
+                    // TODO: num_beams, no_repeat_ngram_size, early_stopping, do_sample, return_full_text
+                }
+
+                let parameters = Parameters {
+                    max_new_tokens: if let Some(Opt::MaxTokens(i)) =
+                        options.get(OptDiscriminants::MaxTokens)
+                    {
+                        Some(*i)
+                    } else {
+                        None
+                    },
```

Comment:
This and the following `if let`s would probably be better expressed with `Option::map` and `Option::and_then`